tracecat/tables/service.py

```diff
@@ -435,30 +538,122 @@ async def insert_row(
         conn = await self.session.connection()
 
         row_data = params.data
         col_map = {c.name: c for c in table.columns}
+        upsert = params.upsert
 
         value_clauses: dict[str, sa.BindParameter] = {}
         cols = []
+
+        table_name_for_logging = table.name
+
         for col, value in row_data.items():
             value_clauses[col] = to_sql_clause(
                 value, col_map[col].name, SqlType(col_map[col].type)
             )
             cols.append(sa.column(self._sanitize_identifier(col)))
 
-        stmt = (
-            sa.insert(sa.table(table.name, *cols, schema=schema_name))
-            .values(**value_clauses)
-            .returning(sa.text("*"))
-        )
-        result = await conn.execute(stmt)
-        await self.session.flush()
-        row = result.mappings().one()
-        return dict(row)
+        if not upsert:
+            stmt = (
+                sa.insert(sa.table(table.name, *cols, schema=schema_name))
+                .values(**value_clauses)
+                .returning(sa.text("*"))
+            )
+        else:
+            # For upsert operations
+            table_obj = sa.table(table.name, *cols, schema=schema_name)
+            pg_stmt = insert(table_obj)
+            pg_stmt = pg_stmt.values(**value_clauses)
+
+            # Determine which columns to use for conflict resolution
+            natural_keys = params.natural_keys
+
+            # Ensure all conflict keys are actually in the data
+            if not all(key in value_clauses for key in natural_keys):
+                raise ValueError("All natural keys must be present in the data")
+
+            # Check if the conflict keys have corresponding columns defined
+            if not natural_keys:
+                raise ValueError("No natural keys specified for upsert")
+
+            # Define what gets updated on conflict
+            update_dict = {
+                col: pg_stmt.excluded[col]
+                for col in value_clauses.keys()
+                if col not in natural_keys  # Don't update the conflict keys
+            }
+
+            try:
+                # Complete the statement with on_conflict_do_update
+                stmt = pg_stmt.on_conflict_do_update(
+                    index_elements=natural_keys, set_=update_dict
+                ).returning(sa.text("*"))
+
+                result = await conn.execute(stmt)
+                await self.session.commit()
+                row = result.mappings().one()
+                return dict(row)
+            except ProgrammingError as e:
+                # Drill down to the root cause
+                original_error = e
+                while (cause := e.__cause__) is not None:
+                    e = cause
+
+                # Check for specific error about missing unique constraint
+                if "there is no unique or exclusion constraint" in str(e):
+                    await self.session.rollback()
```

Comment:
You shouldn't have to explicitly rollback here. I'm pretty sure SQLAlchemy rolls back automatically when we exit the context manager. We have an async context manager https://github.com/TracecatHQ/tracecat/blob/ddf7310d3832c64ea553fd837d9987eca24f2e68/tracecat/db/engine.py#L156 that we use as a fastapi dependency in all router endpoints